### PR TITLE
Adding `get_resources` for resource tracking

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
As part of the resource estimation work this PR adds `get_resources()`. This is the main logic responsible for stepping through a quantum circuit and tracking the resources.
 
**Description of the Change:**
- Added `get_resources()` for tapes, quantum functions and operations

### CODE REVIEW

**Comment by Jaybsoni** [comment]
[sc-76765]

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        return Resources(gate_types=gate_counts_dict)
+
+    res = Resources()  # TODO: Add implementation here!
+    return res
+
+
+@get_resources.register
+def resources_from_qfunc(
+    obj: Callable, gate_set: Set = DefaultGateSet, config: Dict = resource_config
+) -> Resources:
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_tracking.py:106]: Nice catch, added

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        if isinstance(op, ResourceOperator):
+            cmp_rep_ops.append(op.resource_rep_from_op())
+
+        else:
+            try:
+                cmp_rep_ops.append(_temp_map_func(op).resource_rep_from_op())
+
+            except NotImplementedError:
+                decomp = op.decomposition()
+                cmp_rep_ops.extend(_operations_to_compressed_reps(decomp))
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_tracking.py:241]: Sounds good

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+            _counts_from_compressed_res_op(
+                cmp_rep_op, gate_counts_dict, gate_set=initial_gate_set, config=config
+            )
+
+        # Validation:
+        condensed_gate_counts = defaultdict(int)
+        for sub_cmp_rep, counts in gate_counts_dict.items():
+            _counts_from_compressed_res_op(
+                sub_cmp_rep, condensed_gate_counts, scalar=counts, gate_set=gate_set, config=config
+            )
```
Comment (willjmax) [inline pennylane/labs/resource_estimation/resource_tracking.py:200]: In the first loop we get the counts for every gate in the union of `gate_set` and `_StandardGateSet`, then in the second we decompose the gates from `_StandardGateSet` into gates from `gate_set`, is this correct? Why can't we do this with one loop and just directly decompose into gates from `gate_set`?

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+    @wraps(obj)
+    def wrapper(*args, **kwargs):
+        with AnnotatedQueue() as q:
+            obj(*args, **kwargs)
+
+        operations = tuple(op for op in q.queue if not isinstance(op, MeasurementProcess))
+        compressed_res_ops_lst = _operations_to_compressed_reps(operations)
+
+        initial_gate_set = set.union(gate_set, _StandardGateSet)
```
Comment (willjmax) [inline pennylane/labs/resource_estimation/resource_tracking.py:187]: Why use `_StandardGateSet` instead of `DefaultGateSet`?

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+        _counts_from_compressed_res_op(
+            cmp_rep_op, gate_counts_dict, gate_set=_StandardGateSet, config=config
+        )
+
+    # Validation:
+    condensed_gate_counts = defaultdict(int)
+    for sub_cmp_rep, counts in gate_counts_dict.items():
+        _counts_from_compressed_res_op(
+            sub_cmp_rep, condensed_gate_counts, scalar=counts, gate_set=gate_set, config=config
+        )
```
Comment (willjmax) [inline pennylane/labs/resource_estimation/resource_tracking.py:230]: Same questions as above

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        _counts_from_compressed_res_op(
+            sub_cp_rep, gate_counts_dict, scalar=scalar * counts, gate_set=gate_set, config=config
+        )
+
+    return
+
+
+def _temp_map_func(op: Operation) -> ResourceOperator:
+    """Temp map function"""
+    raise NotImplementedError
```
Comment (willjmax) [inline pennylane/labs/resource_estimation/resource_tracking.py:272]: This is also implemented in the `resource_map_function` branch.

File: pennylane/labs/tests/resource_estimation/test_resource_tracking.py [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+        super().__init__(wires=wires)
+
+    def decomposition(self):
+        decomp = [
+            re.ResourceHadamard(wires=self.wires[0]),
+            re.ResourceHadamard(wires=self.wires[1]),
+            re.ResourceCNOT(wires=[self.wires[0], self.wires[1]]),
+        ]
+
+        return decomp
```
Comment (willjmax) [inline pennylane/labs/tests/resource_estimation/test_resource_tracking.py:50]: Why is this necessary? Are the resources classes we already have implemented not sufficient?

**Review by willjmax — COMMENTED** [review]
Looks good. Just a few questions about the standard gate set and validation.

File: pennylane/labs/resource_estimation/resource_tracking.py (hunk 3 above)
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_tracking.py:200]: This is an optimization based on the following observation: 

- Most algorithms will be decomposed into ~ O(100,000) or more gates from the set {X, Y, Z, SWAP, RX, RY, RZ, PhaseShift}. <-- Also probably Controlled versions of these gates but I exclude them for now. 
- The conversion from that layer to the final gateset (which is usually) {Hadamard, T, S, CNOT, Toffoli}, is a static function call 

So the idea is instead of calling the static functions O(100,000) times, we instead just count up the gates at that intermediate stage already and then only do a single function call and scale accordingly.

File: pennylane/labs/resource_estimation/resource_tracking.py (hunk 4 above)
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_tracking.py:187]: See the answer below. Essentially we want to track gates at an intermediate stage first before we go all the way down to the leaf nodes.

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+    "SWAP",
+    "CNOT",
+    "S",
+    "T",
+    "Toffoli",
+    "RX",
+    "RY",
+    "RZ",
+    "PhaseShift",
+}
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_tracking.py:46]: Its used in resources_from_qfunc  and resources_from_tape

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+
+resource_config = {
+    "error_rx": 10e-3,
+    "error_ry": 10e-3,
+    "error_rz": 10e-3,
+}
+
+
+@singledispatch
+def get_resources(obj, gate_set: Set = DefaultGateSet, config: Dict = resource_config) -> Resources:
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_tracking.py:66]: Nice catch! Added 👍🏼

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+    obj: Callable, gate_set: Set = DefaultGateSet, config: Dict = resource_config
+) -> Callable:
+    """Get resources from a quantum function which queues operations!"""
+
+    @wraps(obj)
+    def wrapper(*args, **kwargs):
+        with AnnotatedQueue() as q:
+            obj(*args, **kwargs)
+
+        operations = tuple(op for op in q.queue if not isinstance(op, MeasurementProcess))
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_tracking.py:184]: Sure, its probably easier to read as well 👍🏼

File: pennylane/labs/tests/resource_estimation/test_resource_tracking.py (hunk 7 above)
Comment (Jaybsoni) [reply pennylane/labs/tests/resource_estimation/test_resource_tracking.py:50]: This is the base PR that the other PRs pull from so we don't have access to those objects here. Also in general unit testing base classes is done using dummy operations to reduce the coupling between changes in other operations causing tests to fail for other objects.

File: doc/releases/changelog-dev.md [hunk 11]
```diff
@@ -69,8 +69,14 @@
...
   [(#6428)](https://github.com/PennyLaneAI/pennylane/pull/6428)
 
+<<<<<<< HEAD
+* Added `get_resources()` functionality which allows users to extract resources from a quantum function, tape or 
+  resource operation. Additionally added some standard gatesets `DefaultGateSet` to track resources with respect to.
+  [(#6500)](https://github.com/PennyLaneAI/pennylane/pull/6500)
+=======
 * Added `ResourceOperator` classes for QFT and all operators in QFT's decomposition.
   [(#6447)](https://github.com/PennyLaneAI/pennylane/pull/6447)
+>>>>>>> resource_single_qubit
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply doc/releases/changelog-dev.md:79]: Nice catch 👍🏼

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+from pennylane.queuing import AnnotatedQueue
+from pennylane.tape import QuantumScript
+from pennylane.wires import Wires
+
+from .resource_container import CompressedResourceOp, Resources
+from .resource_operator import ResourceOperator
+
+# pylint: disable=dangerous-default-value,protected-access
+
+_StandardGateSet = {
```
Comment (soranjh) [inline pennylane/labs/resource_estimation/resource_tracking.py:31]: Are these dicts needed in different modules? If yes, safer to define them only once to prevent potential mismatch.

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+@singledispatch
+def get_resources(
+    obj, gate_set: Set = DefaultGateSet, config: Dict = resource_config
+) -> Union[Resources, Callable]:
+    r"""Obtain the resources from a quantum circuit or operation in terms of the gates provided
+    in the gate_set.
+
+    Args:
+        obj (Union[Operation, Callable, QuantumScript]): The quantum circuit or operation to obtain resources from.
+        gate_set (Set, optional): A set (str) specifying the names of opertions to track. Defaults to DefaultGateSet.
```
Comment (soranjh) [inline pennylane/labs/resource_estimation/resource_tracking.py:73]:
```suggestion
        gate_set (Set, optional): A set (str) specifying the names of operations to track. Defaults to DefaultGateSet.
```

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+        config (Dict, optional): Additional configurations to specify how resources are tracked. Defaults to resource_config.
+
+    Returns:
+        Resources: The total resources of the quantum circuit.
+
+    Rasies:
```
Comment (soranjh) [inline pennylane/labs/resource_estimation/resource_tracking.py:79]:
```suggestion
    Raises:
```

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+        gate_types:
+        {'T': 17}
+
+        Finally, we can modify the config values listed in the global :code:`labs.resource_estimation.resource_config`
+        dictionary to have finegrain control of how the resources are computed.
+
+        >>> re.resource_config
+        {'error_rx': 0.01, 'error_ry': 0.01, 'error_rz': 0.01}
+        >>>
+        >>> my_config = copy.copy(re.resource_config)
```
Comment (soranjh) [inline pennylane/labs/resource_estimation/resource_tracking.py:138]: import copy

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+
+
+def _temp_map_func(op: Operation) -> ResourceOperator:
+    """Temp map function"""
+    raise NotImplementedError
+
+
+def _clean_gate_counts(gate_counts: Dict[CompressedResourceOp, int]) -> Dict[str, int]:
+    """Map resources with gate_types made from CompressedResourceOps
+    into one which tracks just strings of operations!
```
Comment (soranjh) [inline pennylane/labs/resource_estimation/resource_tracking.py:278]:
```suggestion
    into one which tracks just strings of operations.
```

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+
+    return clean_gate_counts
+
+
+@qml.QueuingManager.stop_recording()
+def _operations_to_compressed_reps(ops: Iterable[Operation]) -> List[CompressedResourceOp]:
+    """Convert the sequence of operations to a list of compressed resource ops.
+
+    Args:
+        ops (Iterable[Operation]): set of operations to convert.
```
Comment (soranjh) [inline pennylane/labs/resource_estimation/resource_tracking.py:299]:
```suggestion
        ops (Iterable[Operation]): set of operations to convert
```

File: pennylane/labs/resource_estimation/resource_tracking.py [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+
+    Returns:
+        List[CompressedResourceOp]: set of converted compressed resource ops.
```
Comment (soranjh) [inline pennylane/labs/resource_estimation/resource_tracking.py:302]:
```suggestion
        List[CompressedResourceOp]: set of converted compressed resource ops
```

**Review by soranjh — APPROVED** [review]

File: pennylane/labs/resource_estimation/resource_tracking.py (hunk 12 above)
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_tracking.py:31]: They should only be in this module in this file.

File: pennylane/labs/resource_estimation/resource_tracking.py (hunk 15 above)
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_tracking.py:138]: Added 👍🏼
